src/Adapter/Product/ProductDeleter.php

```diff
@@ -0,0 +1,101 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product;
+
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\CannotDeleteProductException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductException;
+use PrestaShop\PrestaShop\Core\Domain\Product\ProductDeleterInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+use PrestaShopException;
+use Product;
+
+/**
+ * Deletes products using legacy object model
+ */
+final class ProductDeleter implements ProductDeleterInterface
+{
+    /**
+     * @var ProductProvider
+     */
+    private $productProvider;
+
+    /**
+     * @param ProductProvider $productProvider
+     */
+    public function __construct(
+        ProductProvider $productProvider
+    ) {
+        $this->productProvider = $productProvider;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete(ProductId $productId): void
+    {
+        $product = $this->productProvider->get($productId);
+
+        if (!$this->deleteProduct($product)) {
+            throw new CannotDeleteProductException(
+                sprintf('Failed to delete product #%d', $product->id),
+                CannotDeleteProductException::FAILED_DELETE
+            );
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function bulkDelete(array $productIds): void
+    {
+        foreach ($productIds as $productId) {
+            $this->deleteProduct($this->productProvider->get($productId));
+        }
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @return bool
+     *
+     * @throws ProductException
+     */
+    private function deleteProduct(Product $product): bool
+    {
+        try {
+            return $product->delete();
```

Comment:
I'm wondering There is no notion of soft delete for Products?

There might not be for now, but we had a discussion about this, it can cause problem to delete a Product associated to an Order, so I think in 178 we should introduce this notion of soft delete
Not in this PR because it has a lot of repercussion in many places of the code, but it should be discussed with @MatShir 

I believe it was discussed in this issue https://github.com/PrestaShop/PrestaShop/issues/20818#issuecomment-688349301

Reply:
hmm. Indeed its a big change and anyways it should be done in separate pr.